modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+/*
+ * Copyright (C) 2004-2013 ZNC, see the NOTICE file for details.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <znc/Chan.h>
+#include <znc/Client.h>
+#include <znc/IRCNetwork.h>
+#include <znc/Modules.h>
+#include <znc/Threads.h>
+#include <znc/User.h>
+#include <znc/Utils.h>
+
+extern "C" {
+#include <libotr/proto.h>
+
+#include <libotr/instag.h>
+#include <libotr/message.h>
+#include <libotr/privkey.h>
+#include <libotr/userstate.h>
+#include <libotr/version.h>
+}
+
+// See http://www.gnupg.org/documentation/manuals/gcrypt/Multi_002dThreading.html
+GCRY_THREAD_OPTION_PTHREAD_IMPL;
+
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <list>
+
+using std::list;
+
+#define PROTOCOL_ID "irc"
+
+/* Due to a bug in libotr-4.0.0, passing OTRL_FRAGMENT_SEND_ALL does not work
+ * because otrInjectMessage callback receives NULL as opdata. This workaround
+ * passes the pointer to the COtrMod instance in a global variable. The bug was
+ * fixed in d748757 thus the workaround shouldn't be needed in libotr > 4.0.0.
+ */
+#if (OTRL_VERSION_MAJOR == 4 && OTRL_VERSION_MINOR == 0 && \
+     OTRL_VERSION_SUB == 0)
+#define INJECT_WORKAROUND_NEEDED
+#endif
+void* inject_workaround_mod;
+
+class COtrGenKeyJob : public CModuleJob {
+  public:
+    COtrGenKeyJob(CModule* pModule)
+        : CModuleJob(pModule, "OtrGenKey", "OTR key generator") {}
+
+    void runThread() override;
+    void runMain() override;
+};
+
+class COtrTimer : public CTimer {
+  public:
+    COtrTimer(CModule* pModule, unsigned int uInterval)
+        : CTimer(pModule, uInterval, /*run forever*/ 0, "OtrTimer",
+                 "OTR message poll") {}
+    virtual ~COtrTimer() {}
+
+  protected:
+    virtual void RunJob();
+};
+
+struct COtrAppData {
+    bool bSmpReply;
+
+    COtrAppData() { bSmpReply = false; }
+
+    static void Add(void* data, ConnContext* context) {
+        context->app_data = new COtrAppData();
+        context->app_data_free = Free;
+    }
+
+    static void Free(void* appdata) {
+        delete static_cast<COtrAppData*>(appdata);
+    }
+};
+
+class COtrMod : public CModule {
+    friend class COtrTimer;
+    friend class COtrGenKeyJob;
+
+  private:
+    static OtrlMessageAppOps m_xOtrOps;
+    OtrlUserState m_pUserState;
+    CString m_sPrivkeyPath;
+    CString m_sFPPath;
+    CString m_sInsTagPath;
+    list<CString> m_Buffer;
```

Comment:
I'll rewrite this to user CBuffer, I probably overlooked that it existed (or didn't exist at the time).